File: bindings/go/pcb.go
package powder

import (
	"encoding/binary"
	"fmt"
	"math"
)

// DataType is one of the four physical column types carried by PCB.
type DataType uint8

const (
	Int64 DataType = iota
	Float64
	Bool
	Utf8
)

func (d DataType) String() string {
	switch d {
	case Int64:
		return "int64"
	case Float64:
		return "float64"
	case Bool:
		return "bool"
	case Utf8:
		return "utf8"
	}
	return fmt.Sprintf("DataType(%d)", uint8(d))
}

const (
	pcbMagic     = 0x31424350 // "PCB1" as a little-endian uint32
	pcbHeaderLen = 24
	pcbColDirLen = 40
)

// Column is a single decoded PCB column. Values are read straight out of the
// little-endian payload; nothing is materialized until you ask for it.
type Column struct {
	name        string
	dtype       DataType
	length      int
	data        []byte
	validityOff int // -1 when the column has no validity bitmap
	buf1Off     int
	buf2Off     int
	utf8Offsets []uint32 // only for Utf8 columns
}

func (c *Column) Name() string   { return c.name }
func (c *Column) Type() DataType { return c.dtype }
func (c *Column) Len() int       { return c.length }

// IsValid reports whether the slot at row holds a value (vs. SQL NULL).
func (c *Column) IsValid(row int) bool {
	if c.validityOff < 0 {
		return true
	}
	b := c.data[c.validityOff+(row>>3)]
	return b&(1<<(uint(row)&7)) != 0
}

func (c *Column) Int64(row int) int64 {
	return int64(binary.LittleEndian.Uint64(c.data[c.buf1Off+row*8:]))
}

func (c *Column) Float64(row int) float64 {
	return math.Float64frombits(binary.LittleEndian.Uint64(c.data[c.buf1Off+row*8:]))
}

func (c *Column) Bool(row int) bool {
	b := c.data[c.buf1Off+(row>>3)]
	return b&(1<<(uint(row)&7)) != 0
}

func (c *Column) String(row int) string {
	start := int(c.utf8Offsets[row])
	end := int(c.utf8Offsets[row+1])
	return string(c.data[c.buf2Off+start : c.buf2Off+end])
}

// Get returns the boxed value at row, or nil for NULL / out of range.
func (c *Column) Get(row int) any {
	if row < 0 || row >= c.length || !c.IsValid(row) {
		return nil
	}
	switch c.dtype {
	case Int64:
		return c.Int64(row)
	case Float64:
		return c.Float64(row)
	case Bool:
		return c.Bool(row)
	case Utf8:
		return c.String(row)
	}
	return nil
}

// Batch is a decoded columnar result set.
type Batch struct {
	numRows int
	columns []*Column
	byName  map[string]*Column
}

func (b *Batch) NumRows() int       { return b.numRows }
func (b *Batch) Columns() []*Column { return b.columns }

// Column returns the named column, or nil.
func (b *Batch) Column(name string) *Column { return b.byName[name] }

// Rows returns a row-oriented view (copies) — handy for small result sets.
func (b *Batch) Rows() []map[string]any {
	rows := make([]map[string]any, b.numRows)
	for r := 0; r < b.numRows; r++ {
		row := make(map[string]any, len(b.columns))
		for _, c := range b.columns {
			row[c.name] = c.Get(r)
		}
		rows[r] = row
	}
	return rows
}

// DecodePCB parses a PCB payload. The batch borrows data; do not mutate it.
func DecodePCB(data []byte) (*Batch, error) {
	if len(data) < pcbHeaderLen {
		return nil, fmt.Errorf("powder: buffer smaller than PCB header")
	}
	if binary.LittleEndian.Uint32(data[0:]) != pcbMagic {
		return nil, fmt.Errorf("powder: not a PCB buffer (bad magic)")
	}
	if v := binary.LittleEndian.Uint16(data[4:]); v != 1 {
		return nil, fmt.Errorf("powder: unsupported PCB version %d", v)
	}
	numColumns := int(binary.LittleEndian.Uint32(data[8:]))
	numRows := int(binary.LittleEndian.Uint32(data[12:]))
	dirOff := int(binary.LittleEndian.Uint32(data[16:]))

	batch := &Batch{
		numRows: numRows,
		columns: make([]*Column, 0, numColumns),
		byName:  make(map[string]*Column, numColumns),
	}
	for i := 0; i < numColumns; i++ {
		d := dirOff + i*pcbColDirLen
		if d+pcbColDirLen > len(data) {
			return nil, fmt.Errorf("powder: truncated PCB directory")
		}
		nameOff := int(binary.LittleEndian.Uint32(data[d:]))
		nameLen := int(binary.LittleEndian.Uint32(data[d+4:]))
		dtype := DataType(data[d+8])
		hasValidity := data[d+9]&1 != 0
		validityOff := int(binary.LittleEndian.Uint32(data[d+12:]))
		buf1Off := int(binary.LittleEndian.Uint32(data[d+20:]))
		buf2Off := int(binary.LittleEndian.Uint32(data[d+28:]))

		if dtype > Utf8 {
			return nil, fmt.Errorf("powder: unsupported PCB type code %d", uint8(dtype))
		}
		col := &Column{
			name:    string(data[nameOff : nameOff+nameLen]),
			dtype:   dtype,
			length:  numRows,
			data:    data,
			buf1Off: buf1Off,
			buf2Off: buf2Off,
		}
		col.validityOff = -1
		if hasValidity {
			col.validityOff = validityOff
		}
		if dtype == Utf8 {
			col.utf8Offsets = make([]uint32, numRows+1)
			for k := 0; k <= numRows; k++ {
				col.utf8Offsets[k] = binary.LittleEndian.Uint32(data[buf1Off+k*4:])
			}
		}
		batch.columns = append(batch.columns, col)
		batch.byName[col.name] = col
	}
	return batch, nil
}

File: bindings/csharp/Powder/Pcb.cs
using System.Buffers.Binary;
using System.Text;

namespace Powder;

/// <summary>One of the four physical column types carried by PCB.</summary>
public enum DataType : byte
{
    Int64 = 0,
    Float64 = 1,
    Bool = 2,
    Utf8 = 3,
}

/// <summary>
/// A single decoded PCB column. Values are read straight out of the
/// little-endian payload; nothing is materialized until you ask for it.
/// </summary>
public sealed class Column
{
    private readonly byte[] _data;
    private readonly int _validityOff; // -1 = no validity bitmap
    private readonly int _buf1Off;
    private readonly int _buf2Off;
    private readonly uint[]? _utf8Offsets;

    internal Column(string name, DataType type, int length, byte[] data,
                    int validityOff, int buf1Off, int buf2Off, uint[]? utf8Offsets)
    {
        Name = name;
        Type = type;
        Length = length;
        _data = data;
        _validityOff = validityOff;
        _buf1Off = buf1Off;
        _buf2Off = buf2Off;
        _utf8Offsets = utf8Offsets;
    }

    public string Name { get; }
    public DataType Type { get; }
    public int Length { get; }

    /// <summary>Whether the slot holds a value (vs. SQL NULL).</summary>
    public bool IsValid(int row)
    {
        if (_validityOff < 0)
        {
            return true;
        }
        return (_data[_validityOff + (row >> 3)] & (1 << (row & 7))) != 0;
    }

    public long GetInt64(int row) =>
        BinaryPrimitives.ReadInt64LittleEndian(_data.AsSpan(_buf1Off + row * 8, 8));

    public double GetDouble(int row) =>
        BinaryPrimitives.ReadDoubleLittleEndian(_data.AsSpan(_buf1Off + row * 8, 8));

    public bool GetBoolean(int row) =>
        (_data[_buf1Off + (row >> 3)] & (1 << (row & 7))) != 0;

    public string GetString(int row)
    {
        var offsets = _utf8Offsets!;
        int start = (int)offsets[row];
        int end = (int)offsets[row + 1];
        return Encoding.UTF8.GetString(_data, _buf2Off + start, end - start);
    }

    /// <summary>Boxed value at <paramref name="row"/>, or null for SQL NULL.</summary>
    public object? Get(int row)
    {
        if (row < 0 || row >= Length || !IsValid(row))
        {
            return null;
        }
        return Type switch
        {
            DataType.Int64 => GetInt64(row),
            DataType.Float64 => GetDouble(row),
            DataType.Bool => GetBoolean(row),
            DataType.Utf8 => GetString(row),
            _ => null,
        };
    }
}

/// <summary>A decoded columnar result set.</summary>
public sealed class Batch
{
    private readonly Dictionary<string, Column> _byName;

    internal Batch(int numRows, List<Column> columns)
    {
        NumRows = numRows;
        Columns = columns;
        _byName = new Dictionary<string, Column>(columns.Count, StringComparer.Ordinal);
        foreach (var c in columns)
        {
            _byName[c.Name] = c;
        }
    }

    public int NumRows { get; }
    public IReadOnlyList<Column> Columns { get; }

    public Column this[string name] =>
        _byName.TryGetValue(name, out var c)
            ? c
            : throw new PowderException($"no such column: {name}");

    /// <summary>Row-oriented view (copies) — convenient for small result sets.</summary>
    public List<Dictionary<string, object?>> ToRows()
    {
        var rows = new List<Dictionary<string, object?>>(NumRows);
        for (int r = 0; r < NumRows; r++)
        {
            var row = new Dictionary<string, object?>(Columns.Count, StringComparer.Ordinal);
            foreach (var c in Columns)
            {
                row[c.Name] = c.Get(r);
            }
            rows.Add(row);
        }
        return rows;
    }

    private const uint Magic = 0x31424350; // "PCB1" little-endian
    private const int ColDirLen = 40;

    /// <summary>Decode a PCB buffer (the wire format in docs/FORMAT.md).</summary>
    public static Batch Decode(byte[] data)
    {
        if (data.Length < 24 || BinaryPrimitives.ReadUInt32LittleEndian(data) != Magic)
        {
            throw new PowderException("not a PCB buffer (bad magic)");
        }
        ushort version = BinaryPrimitives.ReadUInt16LittleEndian(data.AsSpan(4, 2));
        if (version != 1)
        {
            throw new PowderException($"unsupported PCB version {version}");
        }
        int numColumns = (int)BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(8, 4));
        int numRows = (int)BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(12, 4));
        int dirOff = (int)BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(16, 4));

        var columns = new List<Column>(numColumns);
        for (int c = 0; c < numColumns; c++)
        {
            int d = dirOff + c * ColDirLen;
            int nameOff = (int)BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(d, 4));
            int nameLen = (int)BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(d + 4, 4));
            byte typeCode = data[d + 8];
            if (typeCode > 3)
            {
                throw new PowderException($"unsupported PCB type code {typeCode}");
            }
            bool hasValidity = (data[d + 9] & 1) != 0;
            int validityOff = (int)BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(d + 12, 4));
            int buf1Off = (int)BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(d + 20, 4));
            int buf2Off = (int)BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(d + 28, 4));

            string name = Encoding.UTF8.GetString(data, nameOff, nameLen);
            uint[]? utf8Offsets = null;
            if ((DataType)typeCode == DataType.Utf8)
            {
                utf8Offsets = new uint[numRows + 1];
                for (int i = 0; i <= numRows; i++)
                {
                    utf8Offsets[i] =
                        BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(buf1Off + i * 4, 4));
                }
            }
            columns.Add(new Column(name, (DataType)typeCode, numRows, data,
                                   hasValidity ? validityOff : -1, buf1Off, buf2Off, utf8Offsets));
        }
        return new Batch(numRows, columns);
    }
}
